turbopack/crates/turbopack-core/src/module_graph/mod.rs

```diff
@@ -690,8 +702,108 @@ impl SingleModuleGraph {
+        &self,
+        issues: &AutoSet<ResolvedVc<Box<dyn Issue>>>,
+    ) -> Result<FxHashMap<ResolvedVc<Box<dyn Issue>>, Vec<ImportTrace>>> {
+        let issue_paths = issues
+            .iter()
+            .map(|issue| async move { Ok((*issue.file_path().await?).clone()) })
+            .try_join()
+            .await?;
+        let mut file_path_to_traces: FxHashMap<FileSystemPath, Vec<ImportTrace>> =
+            FxHashMap::with_capacity_and_hasher(issue_paths.len(), Default::default());
+        // initialize an empty vec for each path we care about
+        for issue in &issue_paths {
+            file_path_to_traces.entry(issue.clone()).or_default();
+        }
+
+        {
+            let modules = self
+                .modules
+                .iter()
+                .map(|(module, &index)| async move {
+                    Ok(((*module.ident().path().await?).clone(), index))
+                })
+                .try_join()
+                .await?;
+            // Reverse the graph so we can find paths to roots
+            let reversed_graph = Reversed(&self.graph.0);
+            for (path, module_idx) in modules {
+                if let Entry::Occupied(mut entry) = file_path_to_traces.entry(path) {
+                    // compute the path from this index to a root of the graph.
+                    let Some((_, path)) = petgraph::algo::astar(
+                        &reversed_graph,
+                        module_idx,
+                        |n| reversed_graph.neighbors(n).next().is_none(),
+                        // Edge weights
+                        |e| match e.weight() {
+                            // Prefer following normal imports/requires when we can
+                            ChunkingType::Parallel { .. } => 0,
+                            _ => 1,
+                        },
+                        // `astar` can be accelerated with a distance estimation heuristic, as long
+                        // as our estimate is never > the actual distance.
+                        // However we don't have a mechanism, so just
+                        // estimate 0 which essentially makes this behave like
+                        // dijktra's shortest path algorithm.  `petgraph` has an implementation of
+                        // dijkstra's but it doesn't report  paths, just distances.
+                        // NOTE: dijkstra's with integer weights can be accelerated with incredibly
+                        // efficient priority queue structures (basically with only 0 and 1 as
+                        // weights you can use a `VecDeque`!).  However,
+                        // this is unlikely to be a performance concern.
+                        // Furthermore, if computing paths _does_ become a performance concern, the
+                        // solution would be a hand written implementation of dijkstras so we can
+                        // hoist redundant work out of this loop.
```

Comment:
- What about https://docs.rs/petgraph/latest/petgraph/algo/dijkstra/fn.dijkstra.html
- You can actually use a heuristic where the estimate is greater than the actual distance (a "non-admissible" heuristic) as long as you don't require the solution to be optimal. I don't think you have a non-admissible heuristic either though.